File: Example-Mac-Storyboard/Example-Mac-Storyboard/DetailViewController.m
//
//  DetailViewController.m
//  Example-Mac-Storyboard
//
//  Created by Stuart Tevendale on 08/11/2023.
//

#import "DetailViewController.h"
#import "DetailWindowController.h"
#import "PlaybackWindowController.h"
#import "PlaybackManager.h"

#import <AVFoundation/AVFoundation.h>
#import <AVKit/AVKit.h>
#import <AVFAudio/AVFAudio.h>

#import <CocoaUPnP/CocoaUPnP.h>


@interface DetailViewController () <NSTableViewDelegate, NSTableViewDataSource>
@property (strong, nonatomic) IBOutlet NSTableView *tableView;
@property (strong, nonatomic) NSMutableArray *items;
@property (strong, nonatomic) NSTimer *testTimer;
@property (strong, nonatomic) UPPMediaServerDevice *device;
@property (strong, nonatomic) NSString *objectId;
@property (strong, nonatomic) PlaybackManager *playbackManager;@property (strong, nonatomic) AVPlayer *audioPlayer;

@end

@implementation DetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do view setup here.
    

}

- (void)viewDidAppear {
    
    self.items = [NSMutableArray new];

    // Get the Window Controller
    DetailWindowController *windowController = (DetailWindowController *)[self view].window.windowController;
    
    self.device = windowController.device;
    self.playbackManager = windowController.playbackManager;
    self.objectId = windowController.objectId;
    
    
    [self fetchChildren];
}

- (void)fetchChildren
{
    UPPResponseBlock block = ^(NSDictionary *response, NSError *error) {
        if (response) {
            [self loadResults:response[@"Result"]];
        } else {
            NSLog(@"Error fetching results: %@", error);
        }
    };

    [[self.device contentDirectoryService]
     browseWithObjectID:self.objectId
     browseFlag:BrowseDirectChildren
     filter:@"dc:title,upnp:originalTrackNumber,res,res@duration"
     startingIndex:@0
     requestedCount:@0
     sortCritera:nil
     completion:block];
}

- (void)loadResults:(NSArray *)results
{
    dispatch_async(dispatch_get_main_queue(), ^{
        NSLog(@"%@", results);
        [self.items addObjectsFromArray:results];
        [self.tableView reloadData];
    });
}

- (NSString *)titleForMediaItem:(UPPMediaItem *)item
{
    if ([item.objectClass isEqualToString:@"object.item.audioItem.musicTrack"]) {
        NSString *title = [NSString stringWithFormat:@"%02d - %@",
                           [item.trackNumber intValue], item.itemTitle];
        return title;
    }

    return item.itemTitle;
}



#pragma  mark - NSTableViewDataSource methods

- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return [self.items count];
}

- (NSView *)tableView:(NSTableView *)tableView
   viewForTableColumn:(NSTableColumn *)tableColumn
                  row:(NSInteger)row {
    
    NSTableCellView *cellView = [tableView makeViewWithIdentifier:[tableColumn identifier] owner:self];
    NSString *cellText = @"Cell Text";
    
    UPPMediaItem *item = _items[row];
    NSString *textLabel = [self titleForMediaItem:item];
    NSString *detailTextLabel = [item duration];
    
    if (detailTextLabel) {
        cellText = [NSString stringWithFormat:@"%@ - %@", textLabel, detailTextLabel];
    }
    else {
        cellText = textLabel;
    }

//    NSString *cellText = [NSString stringWithFormat:@"%@ - %@", textLabel, detailTextLabel];
    
    [cellView.textField setStringValue:cellText];
    
    return cellView;
}

- (void)tableView:(NSTableView *)tableView
    didAddRowView:(NSTableRowView *)rowView
           forRow:(NSInteger)row {
    
    NSLog(@"Row view added: %ld", row);
}

- (void)tableViewSelectionDidChange:(NSNotification *)aNotification {
    NSLog(@"Selection Changed...");
    
    UPPMediaItem *item = self.items[_tableView.selectedRow];

    if ([item.objectClass isEqualToString:@"object.item.audioItem.musicTrack"]) {
        if (self.playbackManager) {
            [self.playbackManager playItem:item];
            return;
        }
        else { // Open a new window and playback on local device
            
            NSStoryboard *sb = [NSStoryboard storyboardWithName:@"Main" bundle:nil];
            
            PlaybackWindowController *playBackWindowController = (PlaybackWindowController *)[sb instantiateControllerWithIdentifier:@"playbackWindow"];
            
            playBackWindowController.item = item;
            
            [playBackWindowController showWindow:self];
            
            return;
        }
    }
        
    NSStoryboard *sb = [NSStoryboard storyboardWithName:@"Main" bundle:nil];
    
    DetailWindowController *detailWindowController = (DetailWindowController *)[sb instantiateControllerWithIdentifier:@"detailWindow"];

    detailWindowController.playbackManager = self.playbackManager;
    detailWindowController.device = self.device;
    detailWindowController.objectId = item.objectID;
    detailWindowController.window.title = item.itemTitle;

    [detailWindowController showWindow:self];
}

@end

File: Example-Mac-Swift/Example-Mac-Swift/DetailViewController.swift
//
//  DetailViewController.swift
//  Example-Mac-Swift
//
//  Created by Stuart Tevendale on 09/11/2023.
//

import Cocoa

class DetailViewController: NSViewController, NSTableViewDelegate, NSTableViewDataSource {
    
    var device: UPPMediaServerDevice? = nil
    var objectId: String? = nil
    var playbackManager: PlaybackManager? = nil
    var items: [UPPMediaItem] = []
    @IBOutlet weak var tableView: NSTableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do view setup here.
    }
    
    override func viewDidAppear() {
        let windowController: DetailWindowController = self.view.window?.windowController as! DetailWindowController
        
        self.device = windowController.device
        self.playbackManager = windowController.playbackManager
        self.objectId = windowController.objectId
        
        fetchChildren()
        
        
        
    }
    
    func fetchChildren() {
        let block:UPPResponseBlock = {(response: Optional<Dictionary<AnyHashable, Any>>, error: Optional<any Error> ) in
//        let block:UPPResponseBlock = {(response: NSDictionary, error: NSError ) in
            if response != nil {
                self.loadResults(results: response?["Result"] as! [UPPMediaItem])
            }
            else {
                print("Error fetching results:", error as Any)
            }
        }
        
        self.device!.contentDirectoryService().browse(withObjectID: self.objectId, browseFlag: BrowseDirectChildren, filter: "dc:title,upnp:originalTrackNumber,res,res@duration", startingIndex: 0 as NSNumber, requestedCount: 0 as NSNumber, sortCritera: nil, completion: block)
    }
    
    func loadResults(results: [UPPMediaItem]) {
        DispatchQueue.main.async {
            print(results)
            self.items.append(contentsOf: results)
            self.tableView.reloadData()
        }
    }
    
    func titleFormMediaItem(item: UPPMediaItem) -> String {
        if item.objectClass == "object.item.audioItem.musicTrack" {
            let title:String = (item.trackNumber ?? "-1") + " - " + item.itemTitle
            
            return title
        }
        
        return item.itemTitle
    }
    
    // MARK: - NSTableViewDataSource methods
    func numberOfRows(in tableView: NSTableView) -> Int {
        return self.items.count
    }
    
    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        
        guard let cellView = tableView.makeView(withIdentifier:
                NSUserInterfaceItemIdentifier(rawValue: "AutomaticTableColumnIdentifier.0"),
                owner: self) as? NSTableCellView
            else {
                return nil
            }
        
        var cellText = "Cell Text"
         
        let item: UPPMediaItem = self.items[row] 
        
        let textLabel = titleFormMediaItem(item: item)
        let detailTextLabel = item.duration()
        
        if detailTextLabel != nil {
            cellText = textLabel + " - " + (detailTextLabel ?? "")
        }
        else {
            cellText = textLabel
        }
                
        cellView.textField?.stringValue = cellText
        return cellView
    }
    
    func tableViewSelectionDidChange(_ notification: Notification) {
        print ("Selection changed")
        
        let item: UPPMediaItem = self.items[tableView.selectedRow]
        
        if item.objectClass == "object.item.audioItem.musicTrack" {
            if self.playbackManager != nil {
                playbackManager!.play(item)
                return
            }
            else {
                // Play local
            }
        }
        
        let sb = NSStoryboard(name: "Main", bundle: nil)
        let detailWindowController:DetailWindowController = sb.instantiateController(withIdentifier: "detailWindow") as! DetailWindowController
        detailWindowController.window?.title = item.itemTitle
        detailWindowController.playbackManager = self.playbackManager!
        detailWindowController.device = self.device
        detailWindowController.objectId = item.objectID
        
        detailWindowController.showWindow(self)
    }
}
